control-plane/agents/core/src/core/specs.rs

```diff
@@ -35,6 +37,416 @@ enum SpecError {
+            drop(spec);
+
+            // resource specific validation rules
+            if let Err(error) = Self::validate_destroy(&locked_spec, registry).await {
+                let mut spec = locked_spec.lock().await;
+                spec.set_updating(false);
+                return Err(error);
+            }
+            let mut spec = locked_spec.lock().await;
+
+            // once we've started, there's no going back...
+            spec.set_state(SpecState::Deleting);
+
+            spec.start_destroy_op();
+            let spec_clone = spec.clone();
+            drop(spec);
+
+            if let Err(error) = registry.store_obj(&spec_clone).await {
+                let mut spec = locked_spec.lock().await;
+                spec.clear_op();
+                Err(error)
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Completes a destroy operation by trying to delete the spec from the persistent store.
+    /// If the persistent store operation fails then the spec is marked accordingly and the dirty
+    /// spec reconciler will attempt to update the store when the store is back online.
+    async fn complete_destroy<O, R: Send>(
+        result: Result<R, SvcError>,
+        locked_spec: &Arc<Mutex<Self>>,
+        registry: &Registry,
+    ) -> Result<R, SvcError>
+    where
+        Self: SpecTransaction<O>,
+        Self: StorableObject,
+    {
+        let key = locked_spec.lock().await.key();
+        match result {
+            Ok(val) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.commit_op();
+                let deleted = registry.delete_kv(&key.key()).await;
+                match deleted {
+                    Ok(_) => {
+                        Self::remove_spec(locked_spec, registry).await;
+                        let mut spec = locked_spec.lock().await;
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        let mut spec = locked_spec.lock().await;
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a destroy operation and attempt to log the transaction to the store.
```

Comment:
Copy paste error. This is an **update** operation.

Reply:
dohhhh